snapshotter/demux/cache/evict_test.go

```diff
@@ -0,0 +1,103 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package cache
+
+import (
+	"context"
+	"errors"
+	"net"
+	"testing"
+	"time"
+
+	"github.com/firecracker-microvm/firecracker-containerd/snapshotter/demux/internal"
+	"github.com/firecracker-microvm/firecracker-containerd/snapshotter/demux/proxy"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/goleak"
+)
+
+func getSnapshotter(ctx context.Context, key string) (*proxy.RemoteSnapshotter, error) {
+	return &proxy.RemoteSnapshotter{Snapshotter: &internal.SuccessfulSnapshotter{}}, nil
+}
+
+func TestCacheEvictionOnConnectionFailure(t *testing.T) {
+	defer goleak.VerifyNone(t)
+
+	frequency := 1 * time.Millisecond
+	dial := func(ctx context.Context, namespace string) (net.Conn, error) {
+		return nil, errors.New("Mock snapshotter dial error")
+	}
+	cache := NewRemoteSnapshotterCache(getSnapshotter, EvictOnConnectionFailure(dial, frequency))
+	defer cache.Close()
+
+	_, err := cache.Get(context.Background(), "test")
+	require.NoError(t, err, "Snapshotter not added to cache correctly")
+
+	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+	defer cancel()
+
+	for {
+		select {
+		case <-time.After(10 * time.Millisecond):
+			if len(cache.snapshotters) == 0 {
+				return
+			}
```

Comment:
Saw some data race in this test. You might want to acquire the cache lock before reading here (and other tests in this file).

Reply:
Ty great catch.